docs/madr/decisions/020-kds-redesign.md

```diff
@@ -0,0 +1,179 @@
+
+Resources synchronized from Zone to Global are created by default in a `default` namespace and we should change it in the new version. It seems that namespace current sync mechanism checks if resources synchronized from upstream are in the store, and in case the namespace is differnet it removes the old ones and create the new one in the new namespace. There should be no cleanup required, we need only to keep current sync mechanism to work with incremental xDS.
+
+Tried:
+- upgrade path
+- rollback
+
+It looks that state returns to the previous one and nothing breaks.
+
+### Handshake
+
+Is it possible for the zone control-plane to be newer than the global control-plane?
+
+If the answer is no, then a handshake is not necessary as the client will use the latest available version.
+
+If the answer is yes, then a handshake API may need to be implemented. The GRPC Reflection API allows services to be discovered on the server side, but not on the client side.
+
+The output of the reflection functionality is:
+
+```
+["grpc.reflection.v1alpha.ServerReflection", "kuma.mesh.v1alpha1.GlobalKDSService", "kuma.mesh.v1alpha1.MultiplexService"]```
+```
+
+How would a handshake work?
+
+1. The server exposes the RPC `handshake`.
+2. The client initiates a connection to the server.
+3. The client sends a request that includes the Kuma version and the supported KDS versions:
+```
+ kumaVersion: "2.1.0",
+ supportedKdsVersions: ["0.1.0", "0.2.0"],
+```
+4. Server checks the request and responds with the newest version working for both
+```
+ kdsVersion: "0.2.0"
+```
+5. The client needs to create RPC clients for the specific version.
+
+### Multitenant
+
+Global control-plane should support multitenant, that means we want to achieve that one global control-plane can handle many independent zone control-planes. One approach to accomplish this is by extracting tenant specifics from an authentication token. These details can then be transmitted along with a request to indicate the specific tenant and retrieve solely their deployment information. Another option is to supply this information via a header. It appears that accommodating multitenancy doesn't require additional work on the KDS side.
+
+### Incremental xDS
+
+Incremental xDS in the `DeltaDiscoveryRequest` has a field `resource_names_subscribe` which requires information about resources that we want to subscribe to e.g. policy with name `timeout-1`. If you send `resource_names_subscribe:  "*"` in a `DeltaDiscoveryRequest`, the control plane should respond with all resources of the specified type, not just the ones that have changed since the last update. This is because the `resource_names_subscribe` field is used to filter which resources the control plane should send, and if it is set to `*`, it effectively disables filtering and requests all resources of the specified type. Looks like go-control-plane has no implementation of the client, so it might requires us to implement one.
```

Comment:
Reading "should" is confusing here. Is this how it is now, how it could be or how it is decided it will be via this MADR? For example should the second sentence instead read:

```
If the zone CP includes `resource_names_subscribe:  "*"` in a `DeltaDiscoveryRequest`, the global CP responds with all resources of the specified type, not just the ones that have changed since the last update.
```

Will we have to implement an incremental xDS client or does go-control-plane include one?

Reply:
Go-control-plane has no client for delta xDS and it requires us to make one. We are not using the client provided by go-control-plane for the state of the world so I think we can create one.